src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/ObjectWriter/Dwarf/DwarfInfoWriter.cs

```diff
@@ -0,0 +1,218 @@
+            {
+                _infoSectionWriter.EmitSymbolReference(RelocType.IMAGE_REL_BASED_HIGHLOW, ".debug_info", offset);
+            }
+        }
+
+        public void WriteCodeReference(string sectionSymbolName, long offset = 0)
+        {
+            Debug.Assert(offset >= 0);
+            _infoSectionWriter.EmitSymbolReference(_codeRelocType, sectionSymbolName, offset);
+        }
+
+        public void WriteLineReference(long offset)
+        {
+            Debug.Assert(offset < uint.MaxValue);
+            _infoSectionWriter.EmitSymbolReference(RelocType.IMAGE_REL_BASED_HIGHLOW, ".debug_line", offset);
+        }
+
+        public DwarfExpressionBuilder GetExpressionBuilder()
+        {
+            _expressionBufferWriter.Clear();
+            return new DwarfExpressionBuilder(TargetArchitecture, TargetPointerSize, _expressionBufferWriter);
+        }
+
+        public void WriteExpression(DwarfExpressionBuilder expressionBuilder)
+        {
+            _ = expressionBuilder;
+            WriteULEB128((uint)_expressionBufferWriter.WrittenCount);
+            Write(_expressionBufferWriter.WrittenSpan);
+        }
+
+        public void WriteStartLocationList()
+        {
+            long offset = _locSectionWriter.Position;
+            Debug.Assert(offset < uint.MaxValue);
+            _infoSectionWriter.EmitSymbolReference(RelocType.IMAGE_REL_BASED_HIGHLOW, ".debug_loc", (int)offset);
+        }
+
+        public void WriteLocationListExpression(string methodName, long startOffset, long endOffset, DwarfExpressionBuilder expressionBuilder)
+        {
+            _ = expressionBuilder;
+            _locSectionWriter.EmitSymbolReference(_codeRelocType, methodName, startOffset);
+            _locSectionWriter.EmitSymbolReference(_codeRelocType, methodName, endOffset);
+            _locSectionWriter.WriteLittleEndian<ushort>((ushort)_expressionBufferWriter.WrittenCount);
+            _locSectionWriter.Write(_expressionBufferWriter.WrittenSpan);
+        }
+
+        public void WriteEndLocationList()
+        {
+            _locSectionWriter.Write(stackalloc byte[TargetPointerSize * 2]);
+        }
+
+        public void WriteStartRangeList()
+        {
+            long offset = _rangeSectionWriter.Position;
+            Debug.Assert(offset < uint.MaxValue);
+            _infoSectionWriter.EmitSymbolReference(RelocType.IMAGE_REL_BASED_HIGHLOW, ".debug_ranges", offset);
+        }
+
+        public void WriteRangeListEntry(string symbolName, long startOffset, long endOffset)
+        {
+            _rangeSectionWriter.EmitSymbolReference(_codeRelocType, symbolName, startOffset);
+            _rangeSectionWriter.EmitSymbolReference(_codeRelocType, symbolName, endOffset);
+        }
+
+        public void WriteEndRangeList()
+        {
+            _rangeSectionWriter.Write(stackalloc byte[TargetPointerSize * 2]);
+        }
+
+        public void Dispose()
+        {
+            // Debug.Assert(_dieStack.Count == 0);
```

Comment:
nit: Remove comment unless it's useful to have the assert

Reply:
It's a bit of a double edged sword. The assert is useful to ensure that we do proper stacking of elements (ie. every Start is paired with End). Unforunately, the `Dispose` is also called when an exception is thrown, and in that case the assert would always trigger and make it impossible to see the exception.

I'm leaning towards deleting the line.